src/features/hostapp/hooks/target-hostapp.ts

```diff
@@ -202,58 +211,112 @@ async function getOSReleaseResource(
+						// The OS release has to either have a matching variant,
+						// or have no variant (when it's an invariant/unified release).
+						$or: [{ variant: osVariant }, { variant: '' }],
 					},
 					{
-						$or: [
+						// We still need to be filtering hostApp releases based on the deprecated release_tags,
+						// since the versioning format of balenaOS [2019.10.0.dev, 2022.01.0] was non-semver compliant
+						// and they were not migrated to the release semver fields.
+						$and: [
 							{
 								release_tag: {
 									$any: {
 										$alias: 'rt',
 										$expr: {
 											rt: {
-												value: normalizeVariant(osVariant),
-												tag_key: 'variant',
+												// We can't use the result from balena-semver for this filter,
+												// b/c balena-semver normalizes invalid semvers like
+												// 2022.01.0 to 2022.1.0 and that would no longer
+												// match the tag value.
+												value: normalizeOsVersion(osVersion),
+												tag_key: 'version',
 											},
 										},
 									},
 								},
 							},
 							{
-								$not: {
-									release_tag: {
-										$any: {
-											$alias: 'rt',
-											$expr: {
-												rt: {
-													tag_key: 'variant',
+								// The OS release has to either have a matching variant,
+								// or have no variant release_tag (when it's an invariant/unified release).
+								$or: [
+									{
+										release_tag: {
+											$any: {
+												$alias: 'rt',
+												$expr: {
+													rt: {
+														value: normalizeVariantToLongForm(osVariant),
+														tag_key: 'variant',
+													},
 												},
 											},
 										},
 									},
-								},
+									{
+										$not: {
+											release_tag: {
+												$any: {
+													$alias: 'rt',
+													$expr: {
+														rt: {
+															tag_key: 'variant',
+														},
+													},
+												},
+											},
+										},
+									},
+								],
 							},
 						],
 					},
 				],
 			},
+			$orderby: {
+				// We order the resuts by semver_major DESC so that we always prefer rows
+				// that were matched via the semver fields rather than tags in case of a conflict,
+				// (since versions using only tags would have a 0.0.0 semver_.
+				semver_major: 'desc',
+			},
 		},
-	});
+	})) as [
+		PickDeferred<Release, 'id' | 'belongs_to__application'>?,
+		PickDeferred<Release, 'id' | 'belongs_to__application'>?,
+	];
+	if (releases.length > 1) {
+		ThisShouldNeverHappenError(
+			`Found ${releases.length} hostApp releases matching version ${osVersion} ${osVariant} for device type ${deviceTypeId}.`,
+		);
+	}
+
+	const [release] = releases;
+	return release;
 }
 
-function normalizeVariant(variant: string) {
+function getRevisionFromSemver(parsedOsVersion: SemVer) {
+	for (const buildPart of parsedOsVersion.build) {
+		const match = buildPart.match(/^rev(\d+)$/)?.[1];
```

Comment:
```suggestion
	const pattern = /^rev(\d+)$/;
	for (const buildPart of parsedOsVersion.build) {
		const match = buildPart.match(pattern)?.[1];
```
Define the RegExp outside of the loop for better performance. May even be worth defining it outside of the function itself to completely reduce the number of times it ends up being compiled.